wave_optimization/include/wave/optimization/factor_graph/variable.hpp

```diff
@@ -0,0 +1,56 @@
+#ifndef WAVE_OPTIMIZATION_FACTOR_GRAPH_VARIABLE_HPP
+#define WAVE_OPTIMIZATION_FACTOR_GRAPH_VARIABLE_HPP
+
+#include "wave/utils/utils.hpp"
+
+namespace wave {
+
+struct Variable {
+    size_t id;
+    size_t size;
+    VecX data;
+
+    Variable() : id{0}, size{0}, data{0} {}
+    Variable(size_t id, size_t size) : id{id}, size{size}, data{size} {}
+
+    std::string toString() {
+        std::ostringstream oss;
+
+        oss << "[";
+        oss << "id: " << this->id << ", ";
+        oss << "size: " << this->size;
+        oss << "]";
+
+        return oss.str();
+    }
+
+    void print() {
+        std::cout << this->toString() << std::endl;
+    }
+
+    bool operator<(const Variable &other) const {
+        if (this->id < other.id) {
+            return true;
+        }
+        return false;
+    }
+
+    VecX &getData() {
+        return data;
+    }
+};
+
+struct PoseVar : Variable {
+    PoseVar(size_t id) : Variable{id, 6} {
```

Comment:
This goes back to our previous discussion. The factor graph has a storage (be it `std::vector`, etc...) of only the base class `FactorVariable` (Variable - old name), if the factor graph were to operate on this base class `FactorVariable` it will not be able to obtain the derived `PoseVar`'s fixed size vector, because `PoseVar` when added into the factorgraph will have been casted down, if we were to cast a `FactorVariable` back to a `PoseVar` sure,we can have a vector of fixed size of 6, but would you agree this complicates design?